solver/src/main/java/org/chocosolver/solver/constraints/nary/binPacking/PropBinPacking.java

```diff
@@ -0,0 +1,409 @@
+/*
+ * This file is part of choco-solver, http://choco-solver.org/
+ *
+ * Copyright (c) 2020, IMT Atlantique. All rights reserved.
+ *
+ * Licensed under the BSD 4-clause license.
+ *
+ * See LICENSE file in the project root for full license information.
+ */
+package org.chocosolver.solver.constraints.nary.binPacking;
+
+import java.util.BitSet;
+import java.util.Comparator;
+import java.util.stream.IntStream;
+import org.chocosolver.memory.IStateInt;
+import org.chocosolver.solver.constraints.Propagator;
+import org.chocosolver.solver.constraints.PropagatorPriority;
+import org.chocosolver.solver.exception.ContradictionException;
+import org.chocosolver.solver.variables.IntVar;
+import org.chocosolver.solver.variables.delta.IIntDeltaMonitor;
+import org.chocosolver.solver.variables.events.IntEventType;
+import org.chocosolver.solver.variables.events.PropagatorEventType;
+import org.chocosolver.util.ESat;
+import org.chocosolver.util.objects.setDataStructures.ISet;
+import org.chocosolver.util.objects.setDataStructures.ISetIterator;
+import org.chocosolver.util.objects.setDataStructures.SetFactory;
+import org.chocosolver.util.objects.setDataStructures.SetType;
+import org.chocosolver.util.procedure.UnaryIntProcedure;
+import org.chocosolver.util.tools.ArrayUtils;
+
+/**
+ * Propagator for a Bin Packing constraint
+ * This propagator is an implementation of filtering rules introduced in the following paper :
+ * Shaw, P. (2004). A Constraint for Bin Packing. In M. Wallace (Ed.), Principles and Practice of Constraint Programming – CP 2004 (pp. 648–662). Springer Berlin Heidelberg.
+ *
+ * @author Arthur Godet <arth.godet@gmail.com>, Jean-Guillaume Fages
+ */
+public class PropBinPacking extends Propagator<IntVar> {
+    private IntVar[] itemBin;
+    protected int[] itemSize;
+    protected IntVar[] binLoad;
+    protected final int offset;
+
+    private final int nbItems;
+    private final int nbAvailableBins;
+
+    protected IIntDeltaMonitor[] monitors;
+    protected ISet[] P;
+    protected ISet[] R;
+    protected IStateInt[] sumR;
+    private IStateInt[] sumP;
+    private BitSet binsToProcess;
+
+    // NoSum parameters and Java variables
+    private boolean useNoSumFiltering;
+    private int sumA;
+    private int sumB;
+    private int sumC;
+    private int k;
+    private int kPrime;
+    private int[] indexSortedBySize;
+    private int[] X;
+    private int xSize;
+
+    private UnaryIntProcedure<Integer> procedure = new UnaryIntProcedure<Integer>() {
+        int item;
+        @Override
+        public UnaryIntProcedure<Integer> set(Integer itemIdx) {
+            item = itemIdx;
+            return this;
+        }
+        @Override
+        public void execute(int bin) throws ContradictionException {
+            bin -= offset;
+            if(bin >= 0 && bin < nbAvailableBins && P[bin].contains(item)) {
+                sumP[bin].add(-itemSize[item]);
```

Comment:
`IStateInt#add` returns the update value, you can use it, instead of getting it later